File: tests/AetherNet.Core.Tests/HeartbeatTests.cs
// SPDX-License-Identifier: MIT

using System.Text.Json;
using AetherNet.Heartbeat;
using AetherNet.Models;
using AetherNet.Protocol;
using AetherNet.Routing;
using Microsoft.Extensions.Logging.Abstractions;
using Xunit;

namespace AetherNet.Core.Tests;

/// <summary>
/// Unit tests for <see cref="HeartbeatService"/> (PacketType.Heartbeat). Uses a fake
/// <see cref="IMeshSender"/> — no transport needed.
/// </summary>
public sealed class HeartbeatTests
{
    private sealed class FakeMeshSender : IMeshSender
    {
        public string LocalUhid { get; set; } = "aether:local:01";
        public List<MeshPacket> Broadcasts { get; } = [];

        public IReadOnlyList<PeerInfo> GetConnectedPeers() => Array.Empty<PeerInfo>();
        public Task<bool> SendAsync(MeshPacket packet, string nextHopUhid, CancellationToken ct = default)
            => Task.FromResult(true);
        public Task<int> BroadcastAsync(MeshPacket packet, CancellationToken ct = default)
        {
            Broadcasts.Add(packet);
            return Task.FromResult(1);
        }
    }

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
    };

    private static HeartbeatService Build(FakeMeshSender sender)
        => new(sender, NullLogger<HeartbeatService>.Instance);

    private static MeshPacket HeartbeatFrom(string source, int sequence, long sentAtMs) => new()
    {
        Type = PacketType.Heartbeat,
        SourceUhid = source,
        DestinationUhid = "*",
        Payload = JsonSerializer.SerializeToUtf8Bytes(
            new HeartbeatPayload { Sequence = sequence, SentAtMs = sentAtMs }, JsonOpts),
    };

    [Theory]
    [InlineData(1, 1700000000000L, "{\"sequence\":1,\"sent_at_ms\":1700000000000}")]
    [InlineData(0, 0L, "{\"sequence\":0,\"sent_at_ms\":0}")]
    public void HeartbeatPayload_SerializesToCanonicalBytes(int sequence, long ms, string expected)
    {
        var payload = new HeartbeatPayload { Sequence = sequence, SentAtMs = ms };
        var json = System.Text.Encoding.UTF8.GetString(JsonSerializer.SerializeToUtf8Bytes(payload, JsonOpts));
        Assert.Equal(expected, json);
    }

    [Fact]
    public async Task Send_BroadcastsHeartbeat_WithIncrementingSequence()
    {
        var sender = new FakeMeshSender();
        var svc = Build(sender);

        await svc.SendHeartbeatAsync();
        await svc.SendHeartbeatAsync();

        Assert.Equal(2, sender.Broadcasts.Count);
        Assert.All(sender.Broadcasts, p => Assert.Equal(PacketType.Heartbeat, p.Type));
        Assert.All(sender.Broadcasts, p => Assert.Equal(1, p.Ttl));

        var first = JsonSerializer.Deserialize<HeartbeatPayload>(sender.Broadcasts[0].Payload, JsonOpts)!;
        var second = JsonSerializer.Deserialize<HeartbeatPayload>(sender.Broadcasts[1].Payload, JsonOpts)!;
        Assert.Equal(1, first.Sequence);
        Assert.Equal(2, second.Sequence);
    }

    [Fact]
    public async Task Handle_RecordsPeerAndRaisesEvent()
    {
        var svc = Build(new FakeMeshSender { LocalUhid = "aether:local:01" });
        PeerLiveness? seen = null;
        svc.PeerSeen += (_, e) => seen = e;

        var ok = await svc.HandleAsync(HeartbeatFrom("aether:peer:aa", 7, 1700000000000L));

        Assert.True(ok);
        Assert.NotNull(seen);
        Assert.Equal("aether:peer:aa", seen!.Uhid);
        Assert.Equal(7, seen.LastSequence);
        Assert.Equal(1700000000000L, seen.LastSentAtMs);

        var known = svc.GetKnownPeers();
        Assert.Single(known);
        Assert.Equal("aether:peer:aa", known[0].Uhid);
    }

    [Fact]
    public async Task Handle_RefreshesExistingPeer()
    {
        var svc = Build(new FakeMeshSender());
        await svc.HandleAsync(HeartbeatFrom("aether:peer:aa", 1, 1000L));
        await svc.HandleAsync(HeartbeatFrom("aether:peer:aa", 2, 2000L));

        var known = svc.GetKnownPeers();
        Assert.Single(known);
        Assert.Equal(2, known[0].LastSequence);
    }

    [Fact]
    public async Task Handle_OwnHeartbeat_IsIgnored()
    {
        var svc = Build(new FakeMeshSender { LocalUhid = "aether:local:01" });
        var ok = await svc.HandleAsync(HeartbeatFrom("aether:local:01", 1, 1000L));
        Assert.False(ok);
        Assert.Empty(svc.GetKnownPeers());
    }

    [Fact]
    public async Task Handle_WrongPacketType_ReturnsFalse()
    {
        var svc = Build(new FakeMeshSender());
        var pkt = HeartbeatFrom("aether:peer:aa", 1, 1000L);
        pkt.Type = PacketType.Data;
        Assert.False(await svc.HandleAsync(pkt));
    }

    [Fact]
    public async Task GetLivePeers_IncludesRecentlySeenPeer()
    {
        var svc = Build(new FakeMeshSender());
        await svc.HandleAsync(HeartbeatFrom("aether:peer:aa", 1, 1000L));

        // A just-received heartbeat is live within any generous window.
        var live = svc.GetLivePeers(withinSeconds: 3600);
        Assert.Single(live);
        Assert.Equal("aether:peer:aa", live[0].Uhid);

        // A negative window pushes the recency horizon into the future, so it excludes even a
        // just-seen peer — a deterministic proof the filter filters (no wall-clock race).
        Assert.Empty(svc.GetLivePeers(withinSeconds: -1));
    }
}

File: rust/tests/heartbeat_tests.rs
// SPDX-License-Identifier: MIT
//! Integration tests for the Heartbeat service (PacketType 10).

#[path = "common.rs"]
mod common;

use serde_json::json;
use std::sync::Arc;

use aethernet_protocol::{
    heartbeat::HeartbeatService,
    models::PeerInfo,
    protocol::{MeshPacket, PacketType},
};
use common::FakeMeshSender;

const LOCAL: &str = "aether:local:01";

fn new_svc_for(local: &str) -> (HeartbeatService, Arc<FakeMeshSender>) {
    let sender = FakeMeshSender::new(local);
    let svc = HeartbeatService::new(sender.clone());
    (svc, sender)
}

fn new_svc() -> (HeartbeatService, Arc<FakeMeshSender>) {
    new_svc_for(LOCAL)
}

/// Build an inbound Heartbeat packet from `source` with the given sequence /
/// timestamp. Built independently of the service's private wire struct so the
/// test pins the wire shape.
fn heartbeat_from(source: &str, sequence: i32, sent_at_ms: i64) -> MeshPacket {
    let body = serde_json::to_vec(&json!({
        "sequence": sequence,
        "sent_at_ms": sent_at_ms,
    }))
    .unwrap();
    let mut p = MeshPacket::new(PacketType::Heartbeat, source.to_string());
    p.destination_uhid = "*".to_string();
    p.payload = body;
    p
}

// ─── Send ───────────────────────────────────────────────

#[tokio::test]
async fn send_broadcasts_heartbeat_with_incrementing_sequence() {
    let (svc, sender) = new_svc();

    svc.send_heartbeat().await;
    svc.send_heartbeat().await;

    let bcasts = sender.broadcasts();
    assert_eq!(bcasts.len(), 2);
    for p in &bcasts {
        assert_eq!(p.packet_type, PacketType::Heartbeat);
        assert_eq!(p.ttl, 1);
        assert_eq!(p.source_uhid, LOCAL);
        assert_eq!(p.destination_uhid, "*");
    }

    let first: serde_json::Value = serde_json::from_slice(&bcasts[0].payload).unwrap();
    let second: serde_json::Value = serde_json::from_slice(&bcasts[1].payload).unwrap();
    assert_eq!(first["sequence"], json!(1));
    assert_eq!(second["sequence"], json!(2));
}

#[tokio::test]
async fn send_returns_delivered_peer_count() {
    let (svc, sender) = new_svc();
    // FakeMeshSender.broadcast returns the connected-peer count.
    sender.add_peer(PeerInfo::new("aether:peer:aa".into(), Vec::new()));
    sender.add_peer(PeerInfo::new("aether:peer:bb".into(), Vec::new()));

    let delivered = svc.send_heartbeat().await;
    assert_eq!(delivered, 2);
}

// ─── Handle ─────────────────────────────────────────────

#[tokio::test]
async fn handle_records_peer_and_emits_event() {
    let (svc, _) = new_svc();
    let mut events = svc.subscribe_peer_seen();

    let ok = svc
        .handle(&heartbeat_from("aether:peer:aa", 7, 1_700_000_000_000))
        .await;

    assert!(ok);
    let seen = events.try_recv().expect("expected a peer_seen event");
    assert_eq!(seen.uhid, "aether:peer:aa");
    assert_eq!(seen.last_sequence, 7);
    assert_eq!(seen.last_sent_at_ms, 1_700_000_000_000);

    let known = svc.get_known_peers();
    assert_eq!(known.len(), 1);
    assert_eq!(known[0].uhid, "aether:peer:aa");
}

#[tokio::test]
async fn handle_refreshes_existing_peer() {
    let (svc, _) = new_svc();
    svc.handle(&heartbeat_from("aether:peer:aa", 1, 1000)).await;
    svc.handle(&heartbeat_from("aether:peer:aa", 2, 2000)).await;

    let known = svc.get_known_peers();
    assert_eq!(known.len(), 1, "same peer must not create a second record");
    assert_eq!(known[0].last_sequence, 2);
    assert_eq!(known[0].last_sent_at_ms, 2000);
}

#[tokio::test]
async fn handle_own_heartbeat_is_ignored() {
    let (svc, _) = new_svc();
    let ok = svc.handle(&heartbeat_from(LOCAL, 1, 1000)).await;
    assert!(!ok);
    assert!(svc.get_known_peers().is_empty());
}

#[tokio::test]
async fn handle_wrong_packet_type_returns_false() {
    let (svc, _) = new_svc();
    let mut pkt = heartbeat_from("aether:peer:aa", 1, 1000);
    pkt.packet_type = PacketType::Data;
    assert!(!svc.handle(&pkt).await);
    assert!(svc.get_known_peers().is_empty());
}

#[tokio::test]
async fn handle_malformed_payload_returns_false() {
    let (svc, _) = new_svc();
    let mut pkt = heartbeat_from("aether:peer:aa", 1, 1000);
    pkt.payload = b"not json".to_vec();
    assert!(!svc.handle(&pkt).await);
    assert!(svc.get_known_peers().is_empty());
}

// ─── Live peers ─────────────────────────────────────────

#[tokio::test]
async fn get_live_peers_includes_recently_seen_peer() {
    let (svc, _) = new_svc();
    svc.handle(&heartbeat_from("aether:peer:aa", 1, 1000)).await;

    // A just-received heartbeat is live within any generous window.
    let live = svc.get_live_peers(3600);
    assert_eq!(live.len(), 1);
    assert_eq!(live[0].uhid, "aether:peer:aa");

    // A negative window pushes the recency horizon into the future, so it
    // excludes even a just-seen peer — a deterministic proof the filter filters
    // (no wall-clock race).
    assert!(svc.get_live_peers(-1).is_empty());
}
